packages/navbrand/src/react/index.tsx

```diff
@@ -1,25 +1,29 @@
-import { css, media } from 'glamor'
 import Halo from '@pluralsight/ps-design-system-halo'
-import React from 'react'
-import PropTypes from 'prop-types'
+import { css, media } from 'glamor'
+import React, { MouseEventHandler, ReactNode, RefObject } from 'react'
 
-import filterReactProps from '@pluralsight/ps-design-system-filter-react-props'
+import stylesheet from '../css'
 
-import stylesheet from '../css/index.js'
+// TODO: move to utils
+type AnyTag = keyof JSX.IntrinsicElements
+type PropsOf<Tag extends AnyTag> = JSX.IntrinsicElements[Tag]
```

Comment:
This seems to be the "union of 173 keys" that is mentioned here:

https://blog.andrewbran.ch/polymorphic-react-components/#distributivity-schmistributivity-im-gonna-make-it-work

tsc passes in this code.  But I have some trepidation about this warning:

> TypeScript 3.5 does handle this without crashing by deferring a lot of the work that was happening to simplify conditional types, but do you really want to write components that are just waiting for the right moment to explode?

I'm not sure how this might be manifest, but it seems like a problem:

> Long story short, you’ve just invented a button that cannot be reasoned about within Node’s default heap size. 

Long build times?  Non-performant tooling?  Crashing builds in which we are a dependency?  I'm not sure.  But I feel forewarned.

Reply:
yes totally that is my main concern as well. where i think this is different and ok....is that we only allow a button, a, or div as the element to the NavBrand component. for something that would allow any element, things like the ScreenReaderOnly component or where we allow the `tagName`/`as` props i think we need to explore different solutions like the renderContainer prop

Reply:
now i'm reading the blog article again(i think i'm at 10 times). i still think there could be edge cases that we probably haven't accounted for - mostly in typescript perf. without changing the components api i do think this pr demonstrates the best approach i have. the alternative is to explore new ways to conditionally render tagNames for these polymorphic components

I vote we try this out and see what happens. i'm also going to start another thread in ts-guild

Reply:
Re the `keyof JSX.IntrinsicElements` issue: If we wanted to make the union smaller, could we do `keyof Pick<keyof JSX.IntrinsicElements, 'a'|'button'|'div'>`?

Re the best approach we know of: I still don't know what's better about this approach than the Button component approach we also have.  Will you expound on or demonstrate that, please?  If not better in a substantial way, maybe we don't have to take the downsides here too?

Reply:
gotcha. let me step back real quick and see if it works the same with the attrs types

Reply:
@jaketrent tried using types like in button. ie

```
interface AnchorProps
  extends BaseNavBrandProps,
    React.AnchorHTMLAttributes<HTMLAnchorElement> {
  href: string
  onClick?: MouseEventHandler<HTMLAnchorElement>
}
interface ButtonProps
  extends BaseNavBrandProps,
    React.ButtonHTMLAttributes<HTMLButtonElement> {
  href?: undefined
  onClick: MouseEventHandler<HTMLButtonElement>
}
```

but the compiler is unable to parse the overloads or infer the attrs to use. I think `JSX.IntrinsicElements` is the way to go

Reply:
okey dokey.  Did you ever try decreasing the union size with `Pick<keyof JSX.IntrinsicElements, 'a'|'button'|'div'>`?

Reply:
yeah i did. it doesn't seem to have any effect on the compile time but makes the generic util less useful. at this point i don't think the PropsOf utils is of any concern since it's just a lookup map and doesn't return a large union type - it should be no different than what the react typings are already doing....i think

I will remove the AnyTag type just so it's not accidentally abused

Reply:
If in order to be useful, PropsOf includes `keyof JSX.IntrinsicElement` and then creating a potential perf problem, maybe it shouldn't exist.  

One could write:

```
interface AnchorProps extends BaseNavBrandProps, JSX.IntrinsicElements['a'] {}
```

directly into the NavBrand component file.  This doesn't seem to be overly burdensome.

Am I seeing this right?

Reply:
yeah that makes sense to me. i'll update the republish

Reply:
well....i have no idea why but that doesn't work. i still do think it's a problem because we're only ever allowing up to 3 of the elements. we avoid the problem from the blog post where they are allowing all elements

> Ok, fine. Let’s say you work out a way around that assignability error, and you’re ready to replace 'a' | 'button' with all keyof JSX.IntrinsicElements.
> …and, congratulations, you’ve crashed TypeScript 3.4!

we haven't gone that far...yet 😉

Reply:
Huh, weird.  Ok, well, it is what it it is for now, I suppose.  

Thanks for talking through it.